File: implementations/ts/src/pred.ts
// The predicate grammar and its evaluator (SPEC-2 §3). Predicates are total, terminating,
// single-delta: they see one delta at a time, never the rest of the set.

import type { Delta, Pointer, Primitive } from "./types.js";

export type Cmp = "eq" | "neq" | "lt" | "lte" | "gt" | "gte" | "prefix" | "inSet";

export type StrMatch =
  | { readonly kind: "exact"; readonly value: string }
  | { readonly kind: "prefix"; readonly value: string }
  | { readonly kind: "inSet"; readonly values: readonly string[] };

export type ValMatch =
  | { readonly kind: "vcmp"; readonly cmp: Cmp; readonly value: Primitive }
  | { readonly kind: "between"; readonly lo: Primitive; readonly hi: Primitive }
  | { readonly kind: "inSet"; readonly values: readonly Primitive[] };

export interface PPred {
  readonly role?: StrMatch;
  readonly targetEntity?: string;
  readonly targetDelta?: string;
  readonly context?: StrMatch;
  readonly targetIsPrimitive?: boolean;
  readonly targetValue?: ValMatch;
}

export type Pred =
  | { readonly kind: "true" }
  | { readonly kind: "false" }
  | {
      readonly kind: "match";
      readonly field: "author" | "timestamp" | "id";
      readonly cmp: Cmp;
      readonly constant: Primitive | readonly Primitive[];
    }
  | { readonly kind: "hasPointer"; readonly ppred: PPred }
  | { readonly kind: "and"; readonly left: Pred; readonly right: Pred }
  | { readonly kind: "or"; readonly left: Pred; readonly right: Pred }
  | { readonly kind: "not"; readonly pred: Pred };

// --- the canonical total order over primitives (ERRATA-2 E3) ------------------------------------

const utf8 = new TextEncoder();

function utf8Compare(a: string, b: string): number {
  const ab = utf8.encode(a);
  const bb = utf8.encode(b);
  const n = Math.min(ab.length, bb.length);
  for (let i = 0; i < n; i++) {
    const d = ab[i]! - bb[i]!;
    if (d !== 0) return d;
  }
  return ab.length - bb.length;
}

function typeRank(v: Primitive): number {
  if (typeof v === "boolean") return 0;
  if (typeof v === "number") return 1;
  return 2;
}

// Type rank first (bool < number < string), then value; strings by NFC UTF-8 bytes.
export function comparePrimitives(a: Primitive, b: Primitive): number {
  const ra = typeRank(a);
  const rb = typeRank(b);
  if (ra !== rb) return ra - rb;
  if (typeof a === "boolean") return (a ? 1 : 0) - ((b as boolean) ? 1 : 0);
  if (typeof a === "number") {
    const bn = b as number;
    return a < bn ? -1 : a > bn ? 1 : 0;
  }
  return utf8Compare(a, b as string);
}

function compareWith(
  cmp: Cmp,
  subject: Primitive,
  constant: Primitive | readonly Primitive[],
): boolean {
  if (cmp === "inSet") {
    const values = constant as readonly Primitive[];
    return values.some((v) => comparePrimitives(subject, v) === 0);
  }
  if (cmp === "prefix") {
    return (
      typeof subject === "string" && typeof constant === "string" && subject.startsWith(constant)
    );
  }
  const c = comparePrimitives(subject, constant as Primitive);
  switch (cmp) {
    case "eq":
      return c === 0;
    case "neq":
      return c !== 0;
    case "lt":
      return c < 0;
    case "lte":
      return c <= 0;
    case "gt":
      return c > 0;
    case "gte":
      return c >= 0;
  }
}

// --- evaluation ----------------------------------------------------------------------------------

export function strMatch(m: StrMatch, s: string): boolean {
  switch (m.kind) {
    case "exact":
      return s === m.value;
    case "prefix":
      return s.startsWith(m.value);
    case "inSet":
      return m.values.includes(s);
  }
}

function valMatch(m: ValMatch, v: Primitive): boolean {
  switch (m.kind) {
    case "vcmp":
      // cmp "inSet" is rejected at parse time (E1) — ValMatch has its own inSet arm.
      return compareWith(m.cmp, v, m.value);
    case "between":
      return comparePrimitives(v, m.lo) >= 0 && comparePrimitives(v, m.hi) <= 0;
    case "inSet":
      return m.values.some((x) => comparePrimitives(v, x) === 0);
  }
}

function pointerMatches(p: PPred, ptr: Pointer): boolean {
  if (p.role !== undefined && !strMatch(p.role, ptr.role)) return false;
  if (p.targetEntity !== undefined) {
    if (ptr.target.kind !== "entity" || ptr.target.entity.id !== p.targetEntity) return false;
  }
  if (p.targetDelta !== undefined) {
    if (ptr.target.kind !== "delta" || ptr.target.deltaRef.delta !== p.targetDelta) return false;
  }
  if (p.context !== undefined) {
    const ctx =
      ptr.target.kind === "entity"
        ? ptr.target.entity.context
        : ptr.target.kind === "delta"
          ? ptr.target.deltaRef.context
          : undefined;
    if (ctx === undefined || !strMatch(p.context, ctx)) return false;
  }
  if (p.targetIsPrimitive !== undefined) {
    if ((ptr.target.kind === "primitive") !== p.targetIsPrimitive) return false;
  }
  if (p.targetValue !== undefined) {
    if (ptr.target.kind !== "primitive" || !valMatch(p.targetValue, ptr.target.value)) return false;
  }
  return true;
}

// Total and terminating: O(|delta|) per evaluation, no data dereference (SPEC-2 §3).
export function evalPred(pred: Pred, delta: Delta): boolean {
  switch (pred.kind) {
    case "true":
      return true;
    case "false":
      return false;
    case "match": {
      const subject: Primitive =
        pred.field === "author"
          ? delta.claims.author
          : pred.field === "timestamp"
            ? delta.claims.timestamp
            : delta.id;
      return compareWith(pred.cmp, subject, pred.constant);
    }
    case "hasPointer":
      return delta.claims.pointers.some((ptr) => pointerMatches(pred.ppred, ptr));
    case "and":
      return evalPred(pred.left, delta) && evalPred(pred.right, delta);
    case "or":
      return evalPred(pred.left, delta) || evalPred(pred.right, delta);
    case "not":
      return !evalPred(pred.pred, delta);
  }
}

File: implementations/rust/src/pred.rs
//! The predicate grammar and its evaluator (SPEC-2 §3). Mirrors ../ts/src/pred.ts.
//! Predicates are total, terminating, single-delta.

use std::cmp::Ordering;

use crate::types::{Delta, Pointer, Primitive, Target};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Cmp {
    Eq,
    Neq,
    Lt,
    Lte,
    Gt,
    Gte,
    Prefix,
    InSet,
}

#[derive(Debug, Clone, PartialEq)]
pub enum StrMatch {
    Exact(String),
    Prefix(String),
    InSet(Vec<String>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum ValMatch {
    Vcmp { cmp: Cmp, value: Primitive },
    Between { lo: Primitive, hi: Primitive },
    InSet(Vec<Primitive>),
}

#[derive(Debug, Clone, Default, PartialEq)]
pub struct PPred {
    pub role: Option<StrMatch>,
    pub target_entity: Option<String>,
    pub target_delta: Option<String>,
    pub context: Option<StrMatch>,
    pub target_is_primitive: Option<bool>,
    pub target_value: Option<ValMatch>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Field {
    Author,
    Timestamp,
    Id,
}

#[derive(Debug, Clone, PartialEq)]
pub enum MatchConst {
    One(Primitive),
    Many(Vec<Primitive>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Pred {
    True,
    False,
    Match {
        field: Field,
        cmp: Cmp,
        constant: MatchConst,
    },
    HasPointer(PPred),
    And(Box<Pred>, Box<Pred>),
    Or(Box<Pred>, Box<Pred>),
    Not(Box<Pred>),
}

// --- the canonical total order over primitives (ERRATA-2 E3) -------------------------------------

fn type_rank(p: &Primitive) -> u8 {
    match p {
        Primitive::Bool(_) => 0,
        Primitive::Num(_) => 1,
        Primitive::Str(_) => 2,
    }
}

/// Type rank first (bool < number < string), then value; strings by NFC UTF-8 bytes
/// (Rust `str` ordering IS bytewise UTF-8 order, and data strings are NFC by validation D11).
pub fn compare_primitives(a: &Primitive, b: &Primitive) -> Ordering {
    let (ra, rb) = (type_rank(a), type_rank(b));
    if ra != rb {
        return ra.cmp(&rb);
    }
    match (a, b) {
        (Primitive::Bool(x), Primitive::Bool(y)) => x.cmp(y),
        (Primitive::Num(x), Primitive::Num(y)) => x
            .partial_cmp(y)
            .expect("numbers are finite by L1 validation"),
        (Primitive::Str(x), Primitive::Str(y)) => x.as_bytes().cmp(y.as_bytes()),
        _ => unreachable!("ranks matched"),
    }
}

fn compare_with(cmp: Cmp, subject: &Primitive, constant: &MatchConst) -> bool {
    match cmp {
        Cmp::InSet => match constant {
            MatchConst::Many(vs) => vs
                .iter()
                .any(|v| compare_primitives(subject, v) == Ordering::Equal),
            MatchConst::One(_) => false, // rejected at parse time (E1)
        },
        Cmp::Prefix => match (subject, constant) {
            (Primitive::Str(s), MatchConst::One(Primitive::Str(p))) => s.starts_with(p.as_str()),
            _ => false,
        },
        _ => {
            let MatchConst::One(c) = constant else {
                return false;
            };
            let o = compare_primitives(subject, c);
            match cmp {
                Cmp::Eq => o == Ordering::Equal,
                Cmp::Neq => o != Ordering::Equal,
                Cmp::Lt => o == Ordering::Less,
                Cmp::Lte => o != Ordering::Greater,
                Cmp::Gt => o == Ordering::Greater,
                Cmp::Gte => o != Ordering::Less,
                Cmp::Prefix | Cmp::InSet => unreachable!("handled above"),
            }
        }
    }
}

// --- evaluation ------------------------------------------------------------------------------------

pub fn str_match(m: &StrMatch, s: &str) -> bool {
    match m {
        StrMatch::Exact(v) => s == v,
        StrMatch::Prefix(v) => s.starts_with(v.as_str()),
        StrMatch::InSet(vs) => vs.iter().any(|v| v == s),
    }
}

fn val_match(m: &ValMatch, v: &Primitive) -> bool {
    match m {
        // cmp InSet is rejected at parse time (E1) — ValMatch has its own InSet arm.
        ValMatch::Vcmp { cmp, value } => compare_with(*cmp, v, &MatchConst::One(value.clone())),
        ValMatch::Between { lo, hi } => {
            compare_primitives(v, lo) != Ordering::Less
                && compare_primitives(v, hi) != Ordering::Greater
        }
        ValMatch::InSet(vs) => vs
            .iter()
            .any(|x| compare_primitives(v, x) == Ordering::Equal),
    }
}

fn pointer_matches(p: &PPred, ptr: &Pointer) -> bool {
    if let Some(m) = &p.role {
        if !str_match(m, &ptr.role) {
            return false;
        }
    }
    if let Some(e) = &p.target_entity {
        match &ptr.target {
            Target::Entity(er) if &er.id == e => {}
            _ => return false,
        }
    }
    if let Some(d) = &p.target_delta {
        match &ptr.target {
            Target::Delta(dr) if &dr.delta == d => {}
            _ => return false,
        }
    }
    if let Some(m) = &p.context {
        let ctx = match &ptr.target {
            Target::Entity(er) => er.context.as_deref(),
            Target::Delta(dr) => dr.context.as_deref(),
            Target::Primitive(_) => None,
        };
        match ctx {
            Some(c) if str_match(m, c) => {}
            _ => return false,
        }
    }
    if let Some(want) = p.target_is_primitive {
        if matches!(ptr.target, Target::Primitive(_)) != want {
            return false;
        }
    }
    if let Some(m) = &p.target_value {
        match &ptr.target {
            Target::Primitive(v) if val_match(m, v) => {}
            _ => return false,
        }
    }
    true
}

/// Total and terminating: O(|delta|) per evaluation, no data dereference (SPEC-2 §3).
pub fn eval_pred(pred: &Pred, delta: &Delta) -> bool {
    match pred {
        Pred::True => true,
        Pred::False => false,
        Pred::Match {
            field,
            cmp,
            constant,
        } => {
            let subject = match field {
                Field::Author => Primitive::Str(delta.claims.author.clone()),
                Field::Timestamp => Primitive::Num(delta.claims.timestamp),
                Field::Id => Primitive::Str(delta.id.clone()),
            };
            compare_with(*cmp, &subject, constant)
        }
        Pred::HasPointer(pp) => delta
            .claims
            .pointers
            .iter()
            .any(|ptr| pointer_matches(pp, ptr)),
        Pred::And(l, r) => eval_pred(l, delta) && eval_pred(r, delta),
        Pred::Or(l, r) => eval_pred(l, delta) || eval_pred(r, delta),
        Pred::Not(p) => !eval_pred(p, delta),
    }
}
